File: modules/api-rest/src/main/java/com/prevengos/plug/api/adapter/inmemory/InMemoryPacientesAdapter.java
package com.prevengos.plug.api.adapter.inmemory;

import com.prevengos.plug.domain.model.Paciente;
import com.prevengos.plug.domain.model.PacienteId;
import com.prevengos.plug.domain.ports.PacientesPort;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class InMemoryPacientesAdapter implements PacientesPort {
    private final Map<PacienteId, Paciente> pacientes = new ConcurrentHashMap<>();

    @Override
    public Paciente registrar(Paciente paciente) {
        Paciente actualizado = new Paciente(
                paciente.id(),
                paciente.nombre(),
                paciente.apellidos(),
                paciente.documentoIdentidad(),
                paciente.fechaNacimiento(),
                paciente.empresa(),
                paciente.centroTrabajo(),
                Instant.now()
        );
        pacientes.put(actualizado.id(), actualizado);
        return actualizado;
    }

    @Override
    public Paciente buscarPorId(PacienteId id) {
        return pacientes.get(id);
    }

    @Override
    public List<Paciente> pacientesActualizadosDesde(Instant desde) {
        return pacientes.values().stream()
                .filter(paciente -> !paciente.actualizadoEn().isBefore(desde))
                .collect(Collectors.toList());
    }
}

File: modules/api-rest/src/main/kotlin/com/prevengos/plug/api/adapter/inmemory/InMemoryPacientesAdapter.kt
package com.prevengos.plug.api.adapter.inmemory

import com.prevengos.plug.domain.model.Paciente
import com.prevengos.plug.domain.model.PacienteId
import com.prevengos.plug.domain.ports.PacientesPort
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

class InMemoryPacientesAdapter : PacientesPort {
    private val pacientes = ConcurrentHashMap<PacienteId, Paciente>()

    override fun registrar(paciente: Paciente): Paciente {
        pacientes[paciente.id] = paciente.copy(actualizadoEn = Instant.now())
        return pacientes.getValue(paciente.id)
    }

    override fun buscarPorId(id: PacienteId): Paciente? = pacientes[id]

    override fun pacientesActualizadosDesde(desde: Instant): List<Paciente> =
        pacientes.values.filter { it.actualizadoEn.isAfter(desde) || it.actualizadoEn == desde }
}
